ikarus/finiteelements/mechanics/materials/hyperelastic/deviatoric/invariantbased.hh

```diff
@@ -0,0 +1,232 @@
+   * \tparam ST the scalartype of the principal stretches
+   * \return SecondDerivative the second derivative
+   */
+  template <typename ST = ScalarType>
+  SecondDerivative<ST> secondDerivativeImpl(const PrincipalStretches<ST>& lambda) const {
+    const Invariants<ST>& invariants = Impl::invariants(lambda);
+    auto dS                          = SecondDerivative<ST>::Zero().eval();
+
+    const auto& mu            = matParameters_;
+    const auto& devInvariants = DeviatoricInvariants<PrincipalStretches<ST>>(lambda);
+    auto [W1, W2]             = devInvariants.value();
+    W1 -= 3.0;
+    W2 -= 3.0;
+    const auto& [dW1dLambda, dW2dLambda]   = devInvariants.firstDerivative();
+    const auto& [ddW1dLambda, ddW2dLambda] = devInvariants.secondDerivative();
+
+    for (const auto p : parameterRange())
+      for (const auto i : dimensionRange())
+        for (const auto j : dimensionRange()) {
+          auto W1pm1p       = powerAndMultiply<ST>(W1, pex_[p] - 1.0, pex_[p]);
+          auto W2qm1q       = powerAndMultiply<ST>(W2, qex_[p] - 1.0, qex_[p]);
+          auto W1pm2pp      = powerAndMultiply<ST>(W1, pex_[p] - 2.0, pex_[p] * (pex_[p] - 1.0));
+          auto W2qm2qq      = powerAndMultiply<ST>(W2, qex_[p] - 2.0, qex_[p] * (qex_[p] - 1.0));
+          auto dW1W2dlambda = dW1dLambda[i] * dW2dLambda[j] + dW1dLambda[j] * dW2dLambda[i];
+          auto factor1      = (W2qm1q * dW1W2dlambda + pow(W2, qex_[p]) * ddW1dLambda(i, j)) * W1pm1p * mu[p];
+          auto factor2      = W2qm1q * ddW2dLambda(i, j) * pow(W1, pex_[p]) * mu[p];
+          auto factor3      = W1pm2pp * pow(W2, qex_[p]) * dW1dLambda[i] * dW1dLambda[j] * mu[p];
+          auto factor4      = W2qm2qq * pow(W1, pex_[p]) * dW2dLambda[i] * dW2dLambda[j] * mu[p];
+          dS(i, j) += factor1 + factor2 + factor3 + factor4;
+          if (i == j) {
+            auto factor5 =
+                mu[p] * (W1pm1p * pow(W2, qex_[p]) * dW1dLambda[i] + pow(W1, pex_[p]) * W2qm1q * dW2dLambda[i]);
+            dS(i, j) -= (1.0 / lambda[i]) * factor5;
+          }
+        }
+    return dS;
+  }
+
+  /**
+   * \brief Rebinds the material to a different scalar type.
+   * \tparam STO The target scalar type.
+   * \return InvariantBasedT<ScalarTypeOther> The rebound InvariantBased material.
+   */
+  template <typename STO>
+  auto rebind() const {
+    return InvariantBasedT<STO, numMatParameters>(pex_, qex_, matParameters_);
+  }
+
+private:
+  Exponents pex_, qex_;
+  MaterialParameters matParameters_;
+
+  inline static constexpr auto parameterRange() { return Dune::range(numMatParameters); }
+  inline static constexpr auto dimensionRange() { return Dune::range(dim); }
+
+  /**
+   * \brief A function that computes \f$ x^p * m \f$.
+   *
+   * \details In the undeformed configuration, all principal stretches are equal to one. In this scenario, the
+   * deviatoric invariants W1 and W2 are zero. For any positive integer n, 0^{-n} is inf. Multiplying this with 0.0
```

Comment:
This case actually never occurs because you said `std::abs(x) < epsilon and p < 0)` is never triggered? I'm again a bit confused. Otherwise this is fine even though I dont really understand what inputs can actually be expected.